Src/Base/AMReX_IntVect.H

```diff
@@ -34,196 +35,235 @@ int coarsen (int i, int ratio) noexcept
-#if (AMREX_SPACEDIM == 1)
-        return Dim3{vect[0],0,0};
-#elif (AMREX_SPACEDIM == 2)
-        return Dim3{vect[0],vect[1],0};
-#else
-        return Dim3{vect[0],vect[1],vect[2]};
-#endif
+        if constexpr (dim == 1) {
+            return Dim3{vect[0],0,0};
+        } else if constexpr (dim == 2) {
+            return Dim3{vect[0],vect[1],0};
+        } else {
+            return Dim3{vect[0],vect[1],vect[2]};
+        }
     }
 
 #if __cplusplus >= 201402L
     template< typename T = int >
     AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-    Array<T, AMREX_SPACEDIM>
+    Array<T, dim>
     toArray () const noexcept {
-        return Array<T, AMREX_SPACEDIM>{
-            AMREX_D_DECL(T(vect[0]), T(vect[1]), T(vect[2]))
-        };
+        Array<T, dim> ret {};
+        for (int i=0; i<dim; ++i) {
+            ret[i] = T(vect[i]);
+        }
+        return ret;
     }
 #endif
 
   ///
   /**
-     Sum of all components of this IntVect.
+     Sum of all components of this IntVectND.
   */
     [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
     int sum () const noexcept
     {
-        int retval = AMREX_D_TERM((*this)[0], + (*this)[1], +(*this)[2]);
+        int retval = vect[0];
+        for (int i=1; i<dim; ++i) {
+            retval += vect[i];
+        }
         return retval;
     }
 
     //! maximum (no absolute values) value
     [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
     int max () const noexcept
     {
-#if (AMREX_SPACEDIM == 1)
-        return vect[0];
-#elif (AMREX_SPACEDIM == 2)
-        return (vect[0] > vect[1] ? vect[0] : vect[1]);
-#else
-        int r = vect[0] > vect[1] ? vect[0] : vect[1];
-        return (r > vect[2]) ? r : vect[2];
-#endif
+        int retval = vect[0];
+        for (int i=1; i<dim; ++i) {
+            retval = retval > vect[i] ? retval : vect[i];
+        }
+        return retval;
     }
 
     //! minimum (no absolute values) value
     [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
     int min () const noexcept
     {
-#if (AMREX_SPACEDIM == 1)
-        return vect[0];
-#elif (AMREX_SPACEDIM == 2)
-        return (vect[0] < vect[1] ? vect[0] : vect[1]);
-#else
-        int r = vect[0] < vect[1] ? vect[0] : vect[1];
-        return (r < vect[2]) ? r : vect[2];
-#endif
+        int retval = vect[0];
+        for (int i=1; i<dim; ++i) {
+            retval = retval < vect[i] ? retval : vect[i];
+        }
+        return retval;
     }
 
     //return coordinate with largest value
     [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
     int maxDir(bool a_doAbsValue) const noexcept;
 
-    //! Returns a reference to the i'th coordinate of the IntVect.
-    [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-    int& operator[] (int i) noexcept { BL_ASSERT(i>=0 && i < AMREX_SPACEDIM); return vect[i]; }
+    //! Returns a reference to the i'th coordinate of the IntVectND.
+    [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE constexpr
```

Comment:
Because of the use of `BL_ASSERT`, are we allowed to mark this as constexpr?

Reply:
Probably since `amrex::Assert` is not constexpr. However regular `assert` apparently is since C++14 so maybe we can use that or make `amrex::Assert` constexpr.

Reply:
`static_assert(IntVect::TheNodeVector()[0] == 1);` compiles in debug mode.

`static_assert(IntVect::TheNodeVector()[3] == 1);` in debug mode fails with
```
main.cpp:10:49:   in ‘constexpr’ expansion of ‘amrex::IntVectND<3>::TheNodeVector().amrex::IntVectND<3>::operator[](3)’
/home/wqzhang/mygitrepo/amrex/Src/Base/AMReX_BLassert.H:45:54: error: call to non-‘constexpr’ function ‘void amrex::Assert(const char*, const char*, int, const char*)’
   45 | #define    BL_ASSERT(EX) (EX)?((void)0):amrex::Assert( # EX , __FILE__, __LINE__)
      |                                         ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/wqzhang/mygitrepo/amrex/Src/Base/AMReX_IntVect.H:229:40: note: in expansion of macro ‘BL_ASSERT’
  229 |     int& operator[] (int i) noexcept { BL_ASSERT(i>=0 && i < dim); return vect[i]; }
      |                                        ^~~~~~~~~
```

Non-debug build fails with
```
   10 |         static_assert(IntVect::TheNodeVector()[3] == 1);
main.cpp:10:51: error: array subscript value ‘3’ is outside the bounds of array type ‘int [3]’
```

Maybe we can still leave it as constexpr.

Reply:
I see the same thing:
In a test it seems to kind of work as-is. When compiling in debug mode and using IntVect in a constexpr context, the index is known at compile time. If the index is within the bounds it compiles fine, however if the index is outside the bounds, it gives a compiler error
```
/home/asinn/hipace/src/particles/beam/BeamParticleContainer.cpp(508): error: expression must have a constant value
/home/asinn/amrex/Src/Base/AMReX_IntVect.H(229): note #2703-D: cannot call non-constexpr function "amrex::Assert"
/home/asinn/amrex/Src/Base/AMReX.H(167): here
```

Reply:
https://stackoverflow.com/questions/47726503/calling-non-constexpr-function-from-constexpr-allowed-in-some-conditions

Reply:
C++ assert works in the same way
```
<source>: In function 'int main()':
<source>:11:27:   in 'constexpr' expansion of 'func(15)'
<source>:5:5: error: call to non-'constexpr' function 'void __assert_fail(const char*, const char*, unsigned int, const char*)'
    5 |     assert(i < 10);
      |     ^~~~~~
/usr/include/assert.h:69:13: note: 'void __assert_fail(const char*, const char*, unsigned int, const char*)' declared here
   69 | extern void __assert_fail (const char *__assertion, const char *__file,
      |             ^~~~~~~~~~~~~
```